Creation of new package CondFormats/SiPixelTransient and moving some pixel template code there, minor fixes for pixel 2d reconstruction

OzAmram opened the pull request:
This is a follow up pull request to [23263](https://github.com/cms-sw/cmssw/pull/23263).

The most significant change is the creation of the directory CondFormats/SiPixelCommon.
The following files which were previously in RecoLocalTracker/SiPixelRecHits/ are now moved to SiPixelTransient:
interface/SiPixelTemplate.h
interface/SiPixelTemplate2D.h
interface/SiPixelTemplateDefs.h
interface/​SimplePixel.h
interface/SiPixelGenError.h
src/SiPixelTemplate.cc
src/SiPixelTemplate2D.cc
src/SiPixelGenError.cc
Some header files and build files needed to be changed to handle this. 

There are also 3 minor changes for the ClusterRepair code (which is still turned off by default).
The first is a indexing bugfix related to double sized pixels that was producing a 'second bump' in the Y-position residuals. This is now fixed.
The second is a change to make sure the ClusterRepair returns all of the info about the rec hit back to its caller (Probxy, qbin, etc.). There had to be a slight change to CPEGeneric to accommodate this. 
The last change is to not call the actual 2D reconstruction algorithm on very large clusters which are essentially hopeless anyway and were giving bad outputs. Instead we catch these in ClusterRepair and return the best we can. 

This code has been tested on trackingValidation workflow (runTheMatrix 10024.3) with cluster repair off.
It has also been tested on pixel trees with cluster repair turned on in order to confirm that the 2nd peak in the y-residuals is indeed gone. 

@pmaksim1 @perrotta @slava77 @fabiocos @makortel @tvami @tsusa @cmantill @schuetzepaul

Commit:
y-bump bugfix

Commit:
Fixes for large clusters and probxy info

Commit:
Moved files, fixed headers, created build file

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5167

cmsbuild commented:
A new Pull Request was created by  @OzAmram (Oz Amram) for master.

It involves the following packages:

CondFormats/SiPixelCommon
RecoLocalTracker/SiPixelRecHits
SimTracker/SiPixelDigitizer


The following packages do not have a category, yet:

CondFormats/SiPixelCommon
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@perrotta, @cmsbuild, @civanch, @mdhildreth, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @GiacomoSguazzoni, @rovere, @VinInn, @dkotlins, @gpetruc, @ebrondol, @threus, @dgulhan, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented:
Thank you @OzAmram 
Could you please post here the moste relevant validation plots (namely the one showing that the 2nd peak in the y-residuals is gone, and whatever else show noticeable differences wrt the previous code)? Or just a link to a presentation you made at some pixel meeting with them, if you have any

davidlange6 commented:
as to avoid future confusion between CondFormats/SIPixelCommon and CondFormats/SiPixelObjects, I'd suggest changing SiPixelCommon to SiPixelTemplates (assuming the need for a new package is justified)

> On Jun 13, 2018, at 12:28 AM, Oz Amram <notifications@github.com> wrote:
> 
> This is a follow up pull request to 23263.
> 
> The most significant change is the creation of the directory CondFormats/SiPixelCommon.
> The following files which were previously in RecoLocalTracker/SiPixelRecHits/ are now moved to SiPixelCommon:
> interface/SiPixelTemplate.h
> interface/SiPixelTemplate2D.h
> interface/SiPixelTemplateDefs.h
> interface/​SimplePixel.h
> interface/SiPixelGenError.h
> src/SiPixelTemplate.cc
> src/SiPixelTemplate2D.cc
> src/SiPixelGenError.cc
> Some header files and build files needed to be changed to handle this.
> 
> There are also 3 minor changes for the ClusterRepair code (which is still turned off by default).
> The first is a indexing bugfix related to double sized pixels that was producing a 'second bump' in the Y-position residuals. This is now fixed.
> The second is a change to make sure the ClusterRepair returns all of the info about the rec hit back to its caller (Probxy, qbin, etc.). There had to be a slight change to CPEGeneric to accommodate this.
> The last change is to not call the actual 2D reconstruction algorithm on very large clusters which are essentially hopeless anyway and were giving bad outputs. Instead we catch these in ClusterRepair and return the best we can.
> 
> This code has been tested on trackingValidation workflow (runTheMatrix 10024.3) with cluster repair off.
> It has also been tested on pixel trees with cluster repair turned on in order to confirm that the 2nd peak in the y-residuals is indeed gone.
> 
> @pmaksim1 @perrotta @slava77 @fabiocos @makortel @tvami @tsusa @cmantill @schuetzepaul
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/cms-sw/cmssw/pull/23571
> 
> Commit Summary
> 
> 	• y-bump bugfix
> 	• Fixes for large clusters and probxy info
> 	• Moved files, fixed headers, created build file
> File Changes
> 
> 	• A CondFormats/SiPixelCommon/BuildFile.xml (7)
> 	• R CondFormats/SiPixelCommon/interface/SiPixelGenError.h (0)
> 	• R CondFormats/SiPixelCommon/interface/SiPixelTemplate.h (0)
> 	• R CondFormats/SiPixelCommon/interface/SiPixelTemplate2D.h (5)
> 	• R CondFormats/SiPixelCommon/interface/SiPixelTemplateDefs.h (0)
> 	• R CondFormats/SiPixelCommon/interface/SimplePixel.h (0)
> 	• R CondFormats/SiPixelCommon/src/SiPixelGenError.cc (2)
> 	• R CondFormats/SiPixelCommon/src/SiPixelTemplate.cc (4)
> 	• R CondFormats/SiPixelCommon/src/SiPixelTemplate2D.cc (30)
> 	• M RecoLocalTracker/SiPixelRecHits/BuildFile.xml (1)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h (1)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/PixelCPEClusterRepair.h (4)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/PixelCPEGeneric.h (4)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/PixelCPETemplateReco.h (2)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h (4)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco2D.h (4)
> 	• M RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateSplit.h (6)
> 	• M RecoLocalTracker/SiPixelRecHits/src/PixelCPEBase.cc (3)
> 	• M RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc (41)
> 	• M RecoLocalTracker/SiPixelRecHits/src/PixelCPEGeneric.cc (2)
> 	• M SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.h (2)
> Patch Links:
> 
> 	• https://github.com/cms-sw/cmssw/pull/23571.patch
> 	• https://github.com/cms-sw/cmssw/pull/23571.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cmsbuild commented:
-1

Tested at: ae18dfeabca73fe191303b7e477b1083e104b5d3

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28663/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelTemplateDBObjectReader.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelLorentzAngleReader.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelCondObjForHLTBuilder.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelGenErrorDBObjectReader.cc 
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelGenErrorDBObjectReader.cc:12:0:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/poison/RecoLocalTracker/SiPixelRecHits/interface/SiPixelGenError.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
 #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
  ^~~~~
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelCondObjBuilder.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelLorentzAngleDB.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-12-2300/src/CondTools/SiPixel/test/SiPixelFakeCPEGenericErrorParmSourceReader.cc 
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

tvami commented:
@davidlange6 
> I'd suggest changing SiPixelCommon to SiPixelTemplates 
as I understand the reason not to call it SiPixelTemplates is that it also contains the GenError packages.

davidlange6 commented:
if its a separate concept, maybe it (or all) can just go into SiPixelObjects then? [thats where the persistent layer of the same constructs for both the templates and generror already sit afterall]

> On Jun 13, 2018, at 12:38 PM, Tamas Vami <notifications@github.com> wrote:
> 
> @davidlange6
> 
> I'd suggest changing SiPixelCommon to SiPixelTemplates
> as I understand the reason not to call it SiPixelTemplates is that it also contains the GenError packages.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

fabiocos commented:
@davidlange6 I understood that @ggovi was preferring to keep separate the persistent objects to be stored in the DB from the transient objects, I let him comment further in case

davidlange6 commented:
makes some sense (not that cms has very many cond objects with this transient layer..) - maybe the name should reflect what it is - the transient copies of whats is CondFormats/SiPixelObjects?


> On Jun 13, 2018, at 1:18 PM, Fabio Cossutti <notifications@github.com> wrote:
> 
> @davidlange6 I understood that @ggovi was preferring to keep separate the persistent objects to be stored in the DB from the transient objects, I let him comment further in case
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Commit:
Fixed some broken dependencies that didn't show up with cms-checkdeps

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5177

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

OzAmram commented:
Here are the plots showing the y-bump in the residuals being gone 
[residx.pdf](https://github.com/cms-sw/cmssw/files/2098871/residx.pdf)
[residx_higheta.pdf](https://github.com/cms-sw/cmssw/files/2098872/residx_higheta.pdf)
[residy.pdf](https://github.com/cms-sw/cmssw/files/2098873/residy.pdf)
[residy_higheta.pdf](https://github.com/cms-sw/cmssw/files/2098874/residy_higheta.pdf)

And an old one for comparison:
[residy_old.pdf](https://github.com/cms-sw/cmssw/files/2098885/residy_old.pdf)

I also just pushed a new commit that should make everything compile. I had run cms-checkdeps after I made all the changes but I didn't realize that it isn't smart enough to recognize the moved files and find their dependencies so I thought I had gotten them all. I made a new release and modified the files I had moved and then ran checkdeps there to get their dependencies so hopefully I got them all now.

perrotta commented:
please test
(Just to check that it compiles: please run scram build yourself next time and verify that it succeed before committing once again. In any case, I think that some furter iteration will be needed in order to take into account the comments received about which package will finally host the template code)

cmsbuild commented:
The tests are being triggered in jenkins.

OzAmram commented:
Yes I had built with scram myself for the first commits and these ones but the first time I hadn't checked out the package with the build error (CondTools/SiPixel) so it had compiled fine for me. I had also tested it on a fresh install of CMSSW.

pmaksim1 commented:
Regarding the package name, let me add that it will need to contain other
low-level reco code which is now duplicated in FastSim.  So
SiPixelTemplates is quite misleading.

SiPixelObjects is a good name, but the package already has a lot of
completely unrelated code, so I worry that this is going to be confusing to
a lot of people, including us.  A new package will be much better, trust me.

 Thanks!

On Wed, Jun 13, 2018, 12:46 PM Oz Amram <notifications@github.com> wrote:

> Yes I had built with scram myself for the first commits and these ones but
> the first time I hadn't checked out the package with the build error
> (CondTools/SiPixel) so it had compiled fine for me. I had also tested it on
> a fresh install of CMSSW.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/23571#issuecomment-397005589>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEppUamQHRt12MUWATofXykujkVkR28Nks5t8UCQgaJpZM4UlQLh>
> .
>

cmsbuild commented:
+1
Tested at: 01539594ea7b5c8ce73c5878b017f265105a151f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28676/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28676/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902004
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901812
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

OzAmram commented:
So is everyone OK with the new package, CondFormats/SiPixelCommon? Or can we reach a consensus about where would be better?

perrotta commented:
I am personally fine with that name. @davidlange6, do you have strong  
concerns about it, or any better suggestion for indicating that the  
new package contains "the transient copies of whats is  
CondFormats/SiPixelObjects"?
Let also ask @ggovi, just in case.

Moeover: @ggovi, will the new package remain under db responsibility,  
reco, or both? At the moment I would think that the most appropriate  
would be keeping it under both reco and db ownerships, but please let  
us know what do you think about.

Oz Amram <notifications@github.com> ha scritto:

> So is everyone OK with the new package, CondFormats/SiPixelCommon?  
> Or can we reach a consensus about where would be better?
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/23571#issuecomment-397633339

davidlange6 commented:
Its not likely to personally affect me, so no - but will anyone understand the difference between CondFormats/BlahCommon and CondFormats/BlahObjects in 5 years?  Likely not.

If you want suggestions, I make the obvious one: CondFormats/SiPixelTransient


> On Jun 15, 2018, at 5:46 PM, perrotta <notifications@github.com> wrote:
> 
> I am personally fine with that name. @davidlange6, do you have strong 
> concerns about it, or any better suggestion for indicating that the 
> new package contains "the transient copies of whats is 
> CondFormats/SiPixelObjects"?
> Let also ask @ggovi, just in case.
> 
> Moeover: @ggovi, will the new package remain under db responsibility, 
> reco, or both? At the moment I would think that the most appropriate 
> would be keeping it under both reco and db ownerships, but please let 
> us know what do you think about.
> 
> Oz Amram <notifications@github.com> ha scritto:
> 
> > So is everyone OK with the new package, CondFormats/SiPixelCommon? 
> > Or can we reach a consensus about where would be better?
> >
> > --
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub:
> > https://github.com/cms-sw/cmssw/pull/23571#issuecomment-397633339
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ggovi commented:
The usage of 'Common' should reflect the commonalities among different detector/systems. I think David proposal makes sense.
I'm fine that db shares the responsibility with reco.

perrotta commented:
Great, thank you @davidlange6 and @ggovi for your comments and suggestions
@OzAmram @pieterdavid : are you ok with renaming the package ```CondFormats/SiPixelTransient``` as suggested, and integrate the change into this pull request, so that we can continue with it?

pieterdavid commented:
@perrotta 
> @OzAmram @pieterdavid : are you ok with renaming the package CondFormats/SiPixelTransient as suggested, and integrate the change into this pull request, so that we can continue with it?

I think you meant to ask @pmaksim1 instead of me ;-)

OzAmram commented:
Sure fine with me. Not sure if Petar is in email contact right now, but I know his main preference was just that there be a new package, I don't think he will mind this name. I will start working on the changes.

perrotta commented:
Ops, yes, indeed: sorry Pieter for the spam.

Pieter David <notifications@github.com> ha scritto:

> @perrotta
>> @OzAmram @pieterdavid : are you ok with renaming the package  
>> CondFormats/SiPixelTransient as suggested, and integrate the change  
>> into this pull request, so that we can continue with it?
>
> I think you meant to ask @pmaksim1 instead of me ;-)
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/23571#issuecomment-398078924

Commit:
Changed directory of moved pixel code to be CondFormats/SiPixelTransient/

cmsbuild commented:
The code-checks are being triggered in jenkins.

OzAmram commented:
Ok I changed the package to be CondFormats/SiPixelTransient and made the necessary changes. Everything compiles for me.

I guess the description I wrote for this PR is now wrong, can it be changed? Or is that unimportant?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5232

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented:
please @OzAmram change the description and the PR title yourself, as you are the only one allowed to do so (I guess)

cmsbuild commented:
+1
Tested at: 7aa7d7bfc8d57397f736c8ae2980619988a5b674
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28742/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28742/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902013
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901822
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

perrotta commented on CondFormats/SiPixelTransient/src/SiPixelTemplate2D.cc:
```diff
@@ -64,6 +64,7 @@ bool SiPixelTemplate2D::pushfile(int filenum, std::vector< SiPixelTemplateStore2
    // Add template stored in external file numbered filenum to theTemplateStore
    
    // Local variables
+   char c;
```
You are reverting here a commit done by @pmaksim1 in the previous PR : https://github.com/cms-sw/cmssw/commit/fc3c83973262e0007471d8c5710abf0f9a4acc2a#diff-d810f8d340a82d8988d1d72ddfeef082

Please make sure that you work on top of the code which is in the master!
(And in this particular case please revert)

perrotta commented on CondFormats/SiPixelTransient/src/SiPixelTemplate2D.cc:
```diff
@@ -100,10 +101,9 @@ bool SiPixelTemplate2D::pushfile(int filenum, std::vector< SiPixelTemplateStore2
       for (int i=0; (c=in_file.get()) != '\n'; ++i) {
          if(i < 79) {theCurrentTemp.head.title[i] = c;}
-         else       {theCurrentTemp.head.title[79] ='\0';}
+         if(i > 78) {theCurrentTemp.head.title[79] ='\0';}
```
Please revert

perrotta commented on FastSimulation/TrackingRecHitProducer/plugins/BuildFile.xml:
```diff
@@ -30,6 +30,7 @@
 <use   name="PhysicsTools/UtilAlgos"/>
 <use   name="SimDataFormats/TrackingHit"/>
 <use   name="FastSimulation/TrackingRecHitProducer"/>
+<use   name="CondFormats/SiPixelCommon"/>
```
CondFormats/SiPixelTransient

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -307,17 +318,25 @@ PixelCPEClusterRepair::callTempReco2D( DetParam const & theDetParam,
+                       theClusterParam.templYrec_, theClusterParam.templSigmaY_, theClusterParam.probabilityY_,
+                       theClusterParam.templXrec_, theClusterParam.templSigmaX_, theClusterParam.probabilityX_,
+                       theClusterParam.qBin_,
+               speed_, deadpix, zeropix,
```
please indent as in the other lines

Commit:
Fixing reverted changes in Template2D, indenting ClusterRepair, common->transient in fastsim build file

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5256

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

OzAmram commented:
Sorry for the reverted changes. I fixed them and the other other issues with the latest commit

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 09761b32b08d2a27fb20f13be21accf2183883e9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28771/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28771/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902013
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901822
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

perrotta commented:
I wanted to test the modifications introduced by this PR, and therefore I run wf 10824 with PixelRepair enabled, as in https://github.com/cms-sw/cmssw/pull/23263#issuecomment-394072231 with GT 
"102X_upgrade2018_realistic_Candidate_2018_05_20_14_23_53"

The code runs (and this confirms that the migration to a new package is correct), but while looking at the step3 output I notice a huge amount of warning messages as
```
%MSG
%MSG-w OutOfBounds:  TrackProducer:initialStepTracksPreSplitting  20-Jun-2018 11:24:27 CEST Run: 1 Event: 1
Qbin outside the bounds of the quality word. Defaulting to Qbin=0. Qbin = 32637 QualityWord = 0
```
or (quite more often)
```
%MSG
%MSG-w OutOfBounds:  TrackProducer:highPtTripletStepTracks  20-Jun-2018 11:24:28 CEST Run: 1 Event: 1
Prob XY outside the bounds of the quality word. Defaulting to Prob=0. Prob = -0.09999 QualityWord = 0
```

The message about the probability being out of bounds comes from the call in
https://github.com/OzAmram/cmssw/blob/09761b32b08d2a27fb20f13be21accf2183883e9/RecoLocalTracker/SiPixelRecHits/src/PixelCPEBase.cc#L577
and with some little debug I could verify that ```theClusterParam.probabilityX_``` assumes in that cases weird values, as negative numbers, or larger than 1, or even nan.

This smells as being originated by your changes in ```RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc```.

Could you please verify, and fix it if needed?

Commit:
Fixed Probxy issue, correctly applied too large cluster error handling

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5270

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

OzAmram commented:
The probxy issue is actually related to the too large clusters and error handing from the 2D reco. It was a dumb error on my part related to the terrible naming convention that 2D reco is actually called 3d in the code and the regular 1d is called 2d. I also should have caught in my testing beforehand. I apologize. It is fixed now

pmaksim1 commented:
I agree that the naming scheme is terrible.  Discussion with Morris queued for 10.3 :)

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -444,19 +438,27 @@ PixelCPEClusterRepair::callTempReco3D( DetParam const & theDetParam,
+   if(clusterPayload.mrow > 4){
+       // The cluster is too big, the 2D reco will perform horribly.
+       // Better to return immediately in error
+       theClusterParam.ierr = 8;
```
```theClusterParam.ierr2``` , maybe?

pmaksim1 replied:
I concur, ierr2 would make more sense since this replaces the second template fit.  Thanks for catching this.

Commit:
2nd fix for qbin and probxy warnings. All gone now.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5271

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

Commit:
ierr -> ierr2

cmsbuild commented:
The code-checks are being triggered in jenkins.

OzAmram commented:
Thanks for spotting that one, Andrea. Ok everything should be working now. And my own testing shows no more Qbin or probxy warnings.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5272

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -279,9 +282,17 @@ PixelCPEClusterRepair::localPosition(DetParam const & theDetParam, ClusterParam
    theClusterParamBase.probabilityQ_ = theClusterParam.probabilityQ_;
-   theClusterParamBase.probabilityX_ = theClusterParam.templProbXY_;
-   theClusterParamBase.probabilityY_ = 0.;
-   theClusterParamBase.filled_from_2d = true;
```
```
  theClusterParamBase.filled_from_2d = filled_from_2d;
```
(and you can remove the re-assignments below)

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -336,8 +347,8 @@ PixelCPEClusterRepair::callTempReco2D( DetParam const & theDetParam,
-       theClusterParam.templProbY_ = theClusterParam.templProbX_ = theClusterParam.templProbQ_ = 1.0f;
-       theClusterParam.templQbin_ = 0;
+       theClusterParam.templProbY_ = theClusterParam.templProbX_ = theClusterParam.templProbQ_ = 0.;
+       theClusterParam.qBin_ = 0;
```
I'm a bit lost with these "ClusterParam"...
Just to check: there remain still two ```theClusterParam.templQbin_``` in this class: could you please verify and confirm that they were not forgotten?

OzAmram replied:
Yes there is a theClusterParam.templQbin_ but as far as I can tell it is not used at all (?). All the calls to the low level reco code are made just using theClusterParam.qBin_ which is inherited from the base class. I can try to get rid of templQbin_ for clarity if you want.

perrotta replied:
Please get rid of it if not needed, thank you

Commit:
Removed unused variables from ClusterParam Class

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23571/5274

OzAmram commented:
Ok I removed some unnecessary variables from the ClusterParam class, and made the other aesthetic change.

cmsbuild commented:
Pull request #23571 was updated. @perrotta, @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @slava77, @ggovi can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23571/28791/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902133
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901941
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

perrotta commented:
+1
- This PR satisfies the request of disentangling Fast/Full simulation from the reco code
- The implementation corresponds to what was intended; it further integrates a few updates and fixes to the template code
- Tested with running the 10824 wf with PixelRepair enabled, as in https://github.com/cms-sw/cmssw/pull/23263#issuecomment-394072231 with GT "102X_upgrade2018_realistic_Candidate_2018_05_20_14_23_53", and no issue is seen
- With the default configuration no effect is expected in reco output, and none is seen in the jenkins tests

fabiocos commented:
@ggovi @civanch @ssekmen could we quickly converge on the approval of this PR, or do you see residual issues with it? For DB I think we had an agreement, and the changes in simulation are transparent and meant to make SIM independent on RECO

ggovi commented:
+1

civanch commented on CondFormats/SiPixelTransient/src/SiPixelTemplate2D.cc:
```diff
@@ -1149,7 +1149,7 @@ bool SiPixelTemplate2D::xytemp(float xhit, float yhit, bool ydouble[BYM2], bool
+         if(xdouble[m]) {
             //  Combine the x-rows
             for(int n=1; n<BYM3; ++n) {
                dxytempdx[1][m][n] += dxytempdx[1][m+1][n];
```
here and below check of indexes is different is "m+1" always inside array size?

perrotta replied:
Not sure this is the optimal way to deal with it, however ```m``` in ```dxytempdx``` is bound to ```BYM2```
https://github.com/OzAmram/cmssw/blob/94e74992470d17f6fdf3c12caef9712ce034f439/CondFormats/SiPixelTransient/src/SiPixelTemplate2D.cc#L1079
which corresponds to ```TYSIZE+2```
https://github.com/OzAmram/cmssw/blob/94e74992470d17f6fdf3c12caef9712ce034f439/CondFormats/SiPixelTransient/interface/SiPixelTemplateDefs.h#L28

```BXM3``` is ```TYSIZE+1```
https://github.com/OzAmram/cmssw/blob/94e74992470d17f6fdf3c12caef9712ce034f439/CondFormats/SiPixelTransient/interface/SiPixelTemplateDefs.h#L29
and therefore it does not go beyond the edge of the array by construction

civanch replied:
Ok, then extra question: why in this class array are accessed by value in several methods and not by reference?

VinInn replied:
what do you mean with ``array by value"?  These are C-style arrays, if I am not mistaken....

pmaksim1 replied:
@civanch 
I don't know.  The code is from Morris Swartz.  We will need to ask him.  I think the best would be to integrate this PR (since the code runs), so that we can integrate other changes (e.g. in FastSim) that depends on the header change.  Then we can discuss this with Morris and make another PR just to address the sub-optimal performance.  If you'll be at the CMS week, the best would be to sit down with Morris and discuss all these changes in person.

ssekmen commented:
+1

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
